File: ref/relation.c
/*
 * relation.c -- the hard-relation generator Gen for LAS (eprint 2020/845,
 * Section 3, Table 1).  See relation.h for the layering.
 *
 * Gen and KeyGen are DISTINCT algorithms that happen to use the SAME
 * deterministic sampling and matrix arithmetic (paper Section 3, p.7: "the
 * statement-witness generation Gen for R_A runs exactly as KeyGen").  They
 * produce DIFFERENT semantic objects and this file keeps them non-interchangeable:
 *
 *     KeyGen (basesig.c):  r       <- S_1^(n+ell);  t       = A r;
 *                          return (pk, sk)      = (t,       r)
 *     Gen    (this file):  r_prime <- S_1^(n+ell);  t_prime = A r_prime;
 *                          return (Y, witness)  = (t_prime, r_prime)
 *
 * Gen returns a (statement, witness) pair, NOT a key pair.  The honest witness
 * is the paper's r' (in the pair (Y, y) the paper writes y for the generic
 * witness and parses r' := y inside Adapt); we name it r_prime throughout and
 * never `y` (which the paper reuses for the S_gamma masking randomness in Sign /
 * PreSign) and never `s` (reserved for Ext's EXTRACTED witness s = z - z_hat).
 *
 * NO INVENTED HELPERS: the static helpers at the bottom are VERBATIM copies of
 * the las.c / basesig.c local twins of the same name (prefix las_/b_ ->
 * relation_), each a one-to-one twin of a NAMED upstream poly.c/polyvec.c
 * function.  Kept as local copies (not shared) so relation.c links independently
 * of the scheme files and never depends on a layer above it.  Ring-degree loops
 * read LAS_D (= params.h N = 256), the LAS-file convention.
 */
#include <stdint.h>
#include "params.h"
#include "relation.h"   /* shared params (setup.h) + object types (las_types.h) */
#include "poly.h"
#include "randombytes.h"
#include "fips202.h"

/* ---- local helpers, DEFINED AT THE BOTTOM of this file: VERBATIM copies of
 * las.c's local twins (prefix las_ -> relation_), only the ones Gen needs:
 *
 *   relation_rej_S1                              <->  las_rej_S1 (las.c)
 *   relation_poly_uniform_S1                     <->  las_poly_uniform_S1
 *   relation_polyvec_uniform_S1                  <->  las_polyvecm_uniform_S1
 *   relation_polyvecl_ntt                        <->  las_polyvecl_ntt
 *   relation_polyvecl_pointwise_acc_montgomery   <->  las_polyvecl_pointwise_acc_montgomery
 *   relation_polyvec_matrix_pointwise_montgomery <->  las_polyvec_matrix_pointwise_montgomery
 *   relation_polyvecn_reduce                     <->  las_polyvecn_reduce
 *   relation_polyvecn_caddq                      <->  las_polyvecn_caddq
 *   relation_polyvecn_add                        <->  las_polyvecn_add
 *   relation_polyvecn_invntt_tomont              <->  las_polyvecn_invntt_tomont
 */
static unsigned int relation_rej_S1(int32_t *a, unsigned int len, const uint8_t *buf, unsigned int buflen);
static void relation_poly_uniform_S1(poly *a, const uint8_t seed[LAS_SEEDBYTES], uint16_t nonce);
static void relation_polyvec_uniform_S1(poly v[N_PLUS_ELL], const uint8_t seed[LAS_SEEDBYTES], uint16_t nonce);
static void relation_polyvecl_ntt(poly v[ELL]);
static void relation_polyvecl_pointwise_acc_montgomery(poly *w, const poly u[ELL], const poly v[ELL]);
static void relation_polyvec_matrix_pointwise_montgomery(poly t[LAS_N], const poly mat[LAS_N][ELL], const poly v[ELL]);
static void relation_polyvecn_reduce(poly v[LAS_N]);
static void relation_polyvecn_caddq(poly v[LAS_N]);
static void relation_polyvecn_add(poly w[LAS_N], const poly u[LAS_N], const poly v[LAS_N]);
static void relation_polyvecn_invntt_tomont(poly v[LAS_N]);

/*************************************************
* Name:        relation_gen  (Gen -> (Y, r_prime) in R_A, random path)
*
* Description: paper Gen(1^lambda) -> (Y, y) in R_A (Definition 3; Section 3:
*              same sampling and arithmetic as KeyGen).  Draw a fresh seed, then
*              the deterministic body.  Produces a (statement, witness) pair --
*              NOT a key pair.
*
* Returns 0 (success)
**************************************************/
int relation_gen(statement *Y,             /* paper Y = t' = A r' (statement)      */
                 witness *r_prime,         /* paper r' (honest witness, r' <-$ S_1) */
                 const public_params *pp) {  /* paper A: pp = A = [I | A']          */
  uint8_t seed[LAS_SEEDBYTES];   /* PRG seed to sample r' (no paper symbol) */

  randombytes(seed, LAS_SEEDBYTES);
  return relation_gen_seed(Y, r_prime, pp, seed);
}

/*************************************************
* Name:        relation_gen_seed  (deterministic Gen; KAT slot)
*
* Description: same deterministic sampling and arithmetic as base_keygen_seed
*              (basesig.c), but a DISTINCT algorithm producing a DISTINCT object:
*              r' <- S_1^(n+ell); Y = t' = A r'; return the (statement, witness)
*              pair (t', r') -- never (public_key, secret_key).
*
* Returns 0 (success)
**************************************************/
int relation_gen_seed(statement *Y,            /* paper Y = t' = A r' (statement)            */
                      witness *r_prime,        /* paper r' (honest witness, r' <-$ S_1)      */
                      const public_params *pp, /* paper A: pp = A = [I | A'] (public matrix)  */
                      const uint8_t seed[LAS_SEEDBYTES]) {  /* PRG seed to sample r' (no paper symbol) */
  unsigned int j;
  poly r_prime_1_hat[ELL];   /* NTT rep of r'_1 (final ell components of r') */

  /* Gen: r' <-$ S_1^(n+ell)  (same ternary sampler as KeyGen). */
  relation_polyvec_uniform_S1(r_prime->value, seed, 0);

  /* Y = t' = A r' = r'_0 + A' r'_1,  where A = [I_n | A']:
   * r'_0 = first n components (identity block), r'_1 = final ell (meet A'). */
  for(j = 0; j < ELL; ++j)
    r_prime_1_hat[j] = r_prime->value[LAS_N + j];   /* r'_1 = final ell components of r' */
  relation_polyvecl_ntt(r_prime_1_hat);
  relation_polyvec_matrix_pointwise_montgomery(Y->t_prime, pp->a_prime, r_prime_1_hat);  /* A' r'_1 */
  relation_polyvecn_reduce(Y->t_prime);
  relation_polyvecn_invntt_tomont(Y->t_prime);

  relation_polyvecn_add(Y->t_prime, Y->t_prime, r_prime->value);  /* + r'_0 (identity block, top n of r') */
  relation_polyvecn_reduce(Y->t_prime);
  relation_polyvecn_caddq(Y->t_prime);

  /* return (Y, witness) = (t', r') in R_A. */
  return 0;
}

/* ==================== helpers (verbatim las.c twins) ==================== */

/* <-> las_rej_S1 (las.c) <-> b_rej_S1 (basesig.c): 2-bit ternary codes,
 * reject code 3, contiguous stream. */
static unsigned int relation_rej_S1(int32_t *a, unsigned int len,
                                    const uint8_t *buf, unsigned int buflen) {
  unsigned int ctr, pos, s;
  uint8_t byte, v;

  ctr = pos = 0;
  while(ctr < len && pos < buflen) {
    byte = buf[pos++];
    for(s = 0; s < 4 && ctr < len; ++s) {
      v = (byte >> (2*s)) & 3;
      if(v < 3)
        a[ctr++] = (int32_t)v - 1;
    }
  }
  return ctr;
}

/* <-> las_poly_uniform_S1 (las.c). */
static void relation_poly_uniform_S1(poly *a, const uint8_t seed[LAS_SEEDBYTES], uint16_t nonce) {
  unsigned int ctr;
  uint8_t buf[SHAKE256_RATE];
  uint8_t nb[2];
  keccak_state state;

  nb[0] = (uint8_t)nonce;
  nb[1] = (uint8_t)(nonce >> 8);
  shake256_init(&state);
  shake256_absorb(&state, seed, LAS_SEEDBYTES);
  shake256_absorb(&state, nb, 2);
  shake256_finalize(&state);
  shake256_squeezeblocks(buf, 1, &state);

  ctr = relation_rej_S1(a->coeffs, LAS_D, buf, SHAKE256_RATE);

  while(ctr < LAS_D) {
    shake256_squeezeblocks(buf, 1, &state);
    ctr += relation_rej_S1(a->coeffs + ctr, LAS_D - ctr, buf, SHAKE256_RATE);
  }
}

/* <-> las_polyvecm_uniform_S1 (las.c): nonce++ per poly over the (n+ell)-vector. */
static void relation_polyvec_uniform_S1(poly v[N_PLUS_ELL], const uint8_t seed[LAS_SEEDBYTES], uint16_t nonce) {
  unsigned int i;

  for(i = 0; i < N_PLUS_ELL; ++i)
    relation_poly_uniform_S1(&v[i], seed, nonce++);
}

/* <-> las_polyvecl_ntt (las.c): over the ell columns of A'. */
static void relation_polyvecl_ntt(poly v[ELL]) {
  unsigned int i;

  for(i = 0; i < ELL; ++i)
    poly_ntt(&v[i]);
}

/* <-> las_polyvecl_pointwise_acc_montgomery (las.c). */
static void relation_polyvecl_pointwise_acc_montgomery(poly *w, const poly u[ELL],
                                                       const poly v[ELL]) {
  unsigned int i;
  poly t;

  poly_pointwise_montgomery(w, &u[0], &v[0]);
  for(i = 1; i < ELL; ++i) {
    poly_pointwise_montgomery(&t, &u[i], &v[i]);
    poly_add(w, w, &t);
  }
}

/* <-> las_polyvec_matrix_pointwise_montgomery (las.c). */
static void relation_polyvec_matrix_pointwise_montgomery(poly t[LAS_N], const poly mat[LAS_N][ELL],
                                                         const poly v[ELL]) {
  unsigned int i;

  for(i = 0; i < LAS_N; ++i)
    relation_polyvecl_pointwise_acc_montgomery(&t[i], mat[i], v);
}

/* <-> las_polyvecn_reduce (las.c). */
static void relation_polyvecn_reduce(poly v[LAS_N]) {
  unsigned int i;

  for(i = 0; i < LAS_N; ++i)
    poly_reduce(&v[i]);
}

/* <-> las_polyvecn_caddq (las.c). */
static void relation_polyvecn_caddq(poly v[LAS_N]) {
  unsigned int i;

  for(i = 0; i < LAS_N; ++i)
    poly_caddq(&v[i]);
}

/* <-> las_polyvecn_add (las.c): adds the first n components (identity block, r'_0). */
static void relation_polyvecn_add(poly w[LAS_N], const poly u[LAS_N], const poly v[LAS_N]) {
  unsigned int i;

  for(i = 0; i < LAS_N; ++i)
    poly_add(&w[i], &u[i], &v[i]);
}

/* <-> las_polyvecn_invntt_tomont (las.c). */
static void relation_polyvecn_invntt_tomont(poly v[LAS_N]) {
  unsigned int i;

  for(i = 0; i < LAS_N; ++i)
    poly_invntt_tomont(&v[i]);
}

File: rust/fips204-las/src/relation.rs
//! `relation` — the HARD-RELATION layer of LAS (eprint 2020/845): the
//! statement–witness generator `Gen` for the relation
//! `R_A = { (Y, y) = (t, r) : t = A r, ||r||inf <= 1 }` (Table 1), together
//! with the semantic types it owns, [`Statement`] and [`Witness`]
//! (physically defined in `setup.rs` so the codec below both schemes can see
//! them; re-exported here by their owner).  Rust twin of `ref/relation.{c,h}`.
//!
//! Paper model (Section 3, p.7): *"The statement-witness generation Gen for
//! R_A runs exactly as KeyGen."*  So `gen_seed` is the SAME mathematics as
//! `basesig::keygen_seed` — `r' <-$ S_1^(n+ell); t' = A r'` — but it
//! constructs the DISTINCT relation types `(Statement, Witness)`, never
//! `(PublicKey, SecretKey)`: a statement is pk-shaped yet the API keeps the
//! two non-interchangeable (no casts, no aliases).
//!
//! This layer sits between `setup` and the two schemes
//! (`setup -> relation / serialize -> basesig / las`): `las` consumes
//! statements and witnesses; `basesig` never sees them.  Extracted witnesses
//! (Ext's `s = z - ẑ`) live in the EXTENDED relation `R'_A`
//! (`||.||inf <= 2(gamma-kappa)`, the knowledge gap, p.9) — same [`Witness`]
//! type, not assumed ternary.
//!
//! NO INVENTED HELPERS: the two private helpers at the bottom are verbatim
//! twins of the `las.rs` helpers of the same name (which mirror the C helper
//! twins; duplication instead of sharing keeps the layering acyclic, exactly
//! like the C files keep local copies for independent linkability).

#![allow(warnings)]
#![allow(clippy::all, clippy::pedantic, clippy::nursery)]
// Named allows: crate root denies these BY NAME, and named lints outrank the
// `warnings` group regardless of scope depth.
#![allow(
    absolute_paths_not_starting_with_crate,
    dead_code,
    elided_lifetimes_in_paths,
    explicit_outlives_requirements,
    let_underscore_drop,
    macro_use_extern_crate,
    meta_variable_misuse,
    missing_abi,
    missing_docs,
    non_ascii_idents,
    single_use_lifetimes,
    trivial_casts,
    trivial_numeric_casts,
    unreachable_pub,
    unused_extern_crates,
    unused_import_braces,
    unused_lifetimes,
    unused_macro_rules,
    unused_qualifications,
    unused_results,
    variant_size_differences
)]

use core::array::from_fn;

use rand_core::CryptoRngCore;
use sha3::digest::{ExtendableOutput, Update, XofReader};
use sha3::Shake256;
use zeroize::Zeroize;

use crate::helpers::{full_reduce32, mont_reduce, partial_reduce32, to_mont};
use crate::ntt::{inv_ntt, ntt};
use crate::types::{R, R0, T0};

use crate::setup::{PublicParams, D, ELL, N, N_PLUS_ELL, LAS_SEEDBYTES};
// Owner re-export: Statement and Witness belong to THIS layer (physical home
// is las_types.rs, see that module's header).
pub use crate::las_types::{Statement, Witness};

const SHAKE256_RATE: usize = 136;

/// `apply_a` — evaluate the public linear map `Y = A v` on a caller-supplied
/// relation vector.
///
/// ADDITIVE, READ-ONLY, KAT-NEUTRAL (dilithium-msc-project, Stage 2): this
/// introduces no new mathematics — it is a thin public wrapper over the existing
/// private `amul`, which `gen_seed` already uses — and no existing code path is
/// changed, so the pinned KAT digest is unaffected. There is deliberately **no
/// C twin**: `ref/` has no caller, and adding one would put an unused function
/// in the implementation of record.
///
/// WHY IT EXISTS. `A` is expanded into `PublicParams` in the NTT domain and is
/// `pub(crate)`, so an external consumer cannot see the matrix. Stage 2's
/// Groth16 backend must encode the relation `A r = t` as R1CS constraints with
/// `A`'s coefficients as *public constants*, which means it has to recover the
/// matrix. Because `v -> A v` is linear, evaluating it on the unit vectors
/// recovers `A` column by column — this function is what makes that possible
/// without exposing internals or duplicating `expand_a` outside the crate.
///
/// Callers get the canonical `[0,Q)` representative, exactly as `gen_seed` does.
pub fn apply_a(
    pp: &PublicParams,   // paper A: pp = A = [I | A']
    v: &Witness,         // paper relation vector (need NOT be ternary)
) -> Statement {         // paper t' = A v
    Statement(amul(pp, v.as_relation_vector()))
}

/// `gen` — the paper's `Gen(1^lambda) -> (Y, y) in R_A` (Definition 3;
/// Section 3: "runs exactly as KeyGen").  Random path: fresh seed, then the
/// deterministic body.  Math-twin of `basesig::keygen` with relation types
/// constructed.  C twin: `relation_gen` (relation.c).
pub fn gen(
    pp: &PublicParams,             // paper A: pp = A = [I | A']
    rng: &mut impl CryptoRngCore,  // CSPRNG for the seed (no paper symbol)
) -> (Statement, Witness) {        // returns (paper Y, paper y) = (t', r')
    let mut seed = [0u8; LAS_SEEDBYTES]; // PRG seed to sample r' (no paper symbol)
    rng.fill_bytes(&mut seed);
    let out = gen_seed(pp, &seed);
    seed.zeroize(); // the witness is derivable from this seed: wipe
    out
}

/// `gen_seed` — deterministic Gen from an explicit 32-byte seed (reproducible
/// KATs).  Byte-identical mathematics to `basesig::keygen_seed`:
/// `r' <-$ S_1^(n+ell); t' = A r'` — the ONE private generation core the
/// paper prescribes ("Gen runs exactly as KeyGen"), constructing the distinct
/// semantic pair.  C twin: `relation_gen_seed` (relation.c).
pub fn gen_seed(
    pp: &PublicParams,           // paper A: pp = A = [I | A']
    seed: &[u8; LAS_SEEDBYTES],  // PRG seed to sample r' (no paper symbol)
) -> (Statement, Witness) {
    // [PAPER Alg.1] 1-2 (Gen = KeyGen):  r' <-$ S_1^(n+ell)
    let r_prime: [R; N_PLUS_ELL] = from_fn(|j| sample_ternary(seed, j as u16));
    // [PAPER Alg.1] 3 (Gen = KeyGen):    t' = A r'
    let t_prime = amul(pp, &r_prime);
    // [PAPER Alg.1] 4 (Gen = KeyGen):    return (Y, y) = (t', r')
    (Statement(t_prime), Witness::from_relation_vector(r_prime))
}

/* ============================ helpers ============================
 * Verbatim twins of the las.rs helpers of the same name (which carry the
 * full provenance notes down to the upstream/C twins); duplicated here so
 * the relation layer never depends on a scheme module. */

/// w = A*v = v_top + A'*v_bot, with A=[I|A'], A' (pp.a_prime) already in the
/// NTT domain.  Output canonical [0,Q).  Twin of `las.rs amul`.
fn amul(pp: &PublicParams, v: &[R; N_PLUS_ELL]) -> [R; N] {
    // vhat = NTT(v_bot), reduced to (-Q,Q) before use (representative-neutral).
    let vbot: [R; ELL] = from_fn(|j| v[N + j].clone());
    let mut vhat = ntt(&vbot);
    for p in vhat.iter_mut() {
        for x in p.0.iter_mut() {
            *x = partial_reduce32(*x);
        }
    }
    let vm = to_mont(&vhat);

    let mut w: [R; N] = [R0; N];
    for i in 0..N {
        let mut acc = T0;
        for j in 0..ELL {
            for n in 0..D {
                acc.0[n] += mont_reduce(i64::from(pp.a_prime[i][j].0[n]) * i64::from(vm[j].0[n]));
            }
        }
        // Mirrors C poly_reduce(&acc) before the inverse NTT (keeps i32 bounds).
        for x in acc.0.iter_mut() {
            *x = partial_reduce32(*x);
        }
        let lo = inv_ntt(&[acc]); // canonical [0,Q)
        for n in 0..D {
            // identity block + canonicalise (C: poly_add; poly_reduce; poly_caddq)
            w[i].0[n] = full_reduce32(lo[0].0[n] + v[i].0[n]);
        }
    }
    w
}

/// Sample one poly with coefficients uniform in {-1,0,1} (set S_1, ternary).
/// Twin of `las.rs sample_ternary` (2-bit codes, reject 3; contiguous stream).
fn sample_ternary(seed: &[u8; LAS_SEEDBYTES], nonce: u16) -> R {
    let mut h = Shake256::default();
    h.update(seed);
    h.update(&nonce.to_le_bytes());
    let mut rd = h.finalize_xof();
    let mut buf = [0u8; SHAKE256_RATE];
    rd.read(&mut buf);
    let mut pos = 0usize;

    let mut r = R0;
    let mut ctr = 0usize;
    while ctr < D {
        if pos >= SHAKE256_RATE {
            rd.read(&mut buf);
            pos = 0;
        }
        let byte = buf[pos];
        pos += 1;
        let mut s = 0u8;
        while s < 4 && ctr < D {
            let v = (byte >> (2 * s)) & 3;
            if v < 3 {
                r.0[ctr] = i32::from(v) - 1;
                ctr += 1;
            }
            s += 1;
        }
    }
    r
}
